L58-python-async-api.md

```diff
@@ -0,0 +1,1666 @@
+most common use case as an example, the new API will diverge from the old API
+since the creation of `Channel` and `Server` object. New creation methods are
+added to instantiate async `Server` or `Channel` object:
+
+```Python
+import grpc
+server = grpc.aio.server(...)
+channel = grpc.aio.insecure_channel(...)
+channel = grpc.aio.secure_channel(...)
+channel = grpc.aio.intercept_channel(...)
+```
+
+To reduce cognitive burden, this gRFC tries to make the new async API share the same
+parameters as the current API, and keep the usage similar.
+
+### Demo Snippet of Async API
+
+Server side:
+```Python
+class AsyncGreeter(helloworld_pb2_grpc.GreeterServicer):
+
+    async def SayHello(self,
+                       request: helloworld_pb2_grpc.HelloRequest,
+                       context: grpc.aio.ServicerContext
+                ) -> helloworld_pb2_grpc.HelloReply:
+        await asyncio.sleep(1)
+        return helloworld_pb2.HelloReply(message="Hello, %s!" % request.name)
+
+server = grpc.aio.server()
+server.add_insecure_port(":50051")
+helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(), server)
+server.start()
+await server.wait_for_termination()
+```
+
+Client side:
+```Python
+async with grpc.aio.insecure_channel("localhost:50051") as channel:
+    stub = echo_pb2_grpc.EchoStub(channel)
+    response = await stub.Hi(echo_pb2.EchoRequest(message="ping"))
+
+    async for response in stub.StreamingHi(...):
+        process(response)
+```
+
+### New Streaming API
+
+Existing streaming API has usability problem that its logic complexes user
+application. For client side, it requires the request iterator to be defined
+above the line of call, and the consumption of responses to be coded below the
+line of invocation. So when the application wants to send request based on
+received response, it would be quite challenging and usually involves
+multi-threading, which is not optimal in Python world.
+
+For server side, the same problem remains for non-trivial streaming servicer
+handlers. The `yield` syntax constraint the message sending should happen in the
+handler itself, otherwise the user needs to code a dedicated iterator class. For
+larger application, it would be a pain to create such iterator class for each
+streaming handler.
+
+euroelessar@ points out several great reasons for the new streaming API:
+* Enable `with` statement for streaming API, which allows clean resource management;
+* Easier to buffer outbound messages when peer pushes back (e.g. `yield` will
+  stop the execution of current coroutine, but `call.write` can be scheduled in
+  event loop);
+* Enable application to have both pending read and pending write in one
+  coroutine, and react on the first finisher (with iterator syntax this will be
+  quite hard to get it right).
+
+So, this gRFC introduces a new pattern of streaming API that reads/writes
+message to peer with explicit call.
+
+#### Snippet For New Streaming API
```

Comment:
maybe I'm missing something, but the snippet for "new streaming API" looks  way more confusing and less fluent than the "current streaming API", so I'm a bit lost.
Btw, you're talking about "current streaming API" but you're talking about an API that AFAIK doesn't exist yet and it's being proposed in the proposal - hence my confusion.

Reply:
I renamed it into `Reader / Writer` API for the new one, and `Async Iterator` for the existing one. Hope this will reduce the confusion.

Reply:
Much better! Thanks.